File: HashMap/560subarraySumEqualsK.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        counter = collections.Counter()
        res, temp = 0, 0
        counter[0] = 1
        
        for i in nums:
            temp += i
            res += counter[temp-k]
            counter[temp] += 1
        return res

File: HashMap/560SubarraySumEqualsK.cpp
/**
 Solution:
 Need to think more
*/
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int cnt = 0, sum = 0;
        unordered_map<int,int> m;
        m[0] = 1;
        for(int i=0; i<nums.size(); i++){
            sum += nums[i];
            if(m.find(sum-k)!=m.end()){
                cnt += m[sum-k];
            }
            m[sum]++;
        }
        return cnt;
    }
};